docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,318 @@
+through gossiping new interesting `Have`s. This is achieved by maintainig a
+persisted view on the list of identities and their current refs. During
+periodic runs a delta is created which serves as the basis of the announcements.
+In turn that list is made known to actively connected peers via the core
+protocols gossip.
+
+==== Requests
+
+The core protocol implicitly tracks projects after replication, and explicitely
+with the `track` operation for remotes in the context of a `Urn`. To enable the
+same approach as for announcements where offline changes to the storage can be
+picked up at any point the node is running and online, this subroutine will
+maintain a persisted view and periodically (potentially provoked by a request
+over the <<API>>) build the delta for new tracked `Urn`s and remotes in their
+context.
+
+Note: A blocking issue for that is
+https://github.com/radicle-dev/radicle-link/issues/141[#141]
+
+==== Tracking
+
+Given the ability to configure a set of urn, peers, both or anything observed
+from connected peers, the node will automatically track and replicate. This
+strategy is currently in use in the seed node and SHALL be preserved.
+
+== API
+
+=== IPC
+
+All communication with daemon processes SHALL occur over UNIX domain sockets in
+`SOCK_STREAM` mode. The socket files MUST be stored in a directory only
+accessible by the logged-in user (typically `$XDG_RUNTIME_DIR`), and have
+permissions `0700` by default. Per-service socket paths are considered "well
+known".
+
+RPC calls over those sockets use <<cbor, CBOR>> for their payload encoding. As
+incremental decoders are not available on all platforms, CBOR-encoded messages
+shall be prepended by their length in bytes, encoded as a 32-bit unsigned
+integer in network byte order.
+
+RPC messages are wrapped in either a `request` or `response` envelope structure
+as defined below:
+
+[source,cddl]
+----
+request = [
+    request-headers,
+    ? payload: bstr,
+]
+response = ok / error
+ok = [
+    response-headers,
+    ? payload: bstr,
+]
+error = [
+    response-headers,
+    code: uint,
+    ? message: tstr,
+]
+request-headers = {
+    ua: client-id,
+    rq: request-id,
+    ? token: token,
+}
+response-headers = {
+    rq: request-id,
+}
+; Unambiguous, human-readable string identifying the client application. Mainly
+; for diagnostic purposes. Example: "radicle-link-cli/v1.2+deaf"
+client-id: tstr .size (4..16)
+; Request identifier, choosen by the client. The responder includes the
+; client-supplied value in the response, enabling request pipelining.
+;
+; Note that streaming / multi-valued responses may include the same id in
+; several response messages.
+request-id: bstr .size (4..16)
+
+; Placeholder for future one-time-token support.
+token: bstr
+----
+
+
+=== Request/Response payloads
+
+There is a supported set of requests a daemon will answer to which are as follows:
+
+[source,cddl]
+----
+request = discover / memebership / stats
+request-discover = {
```

Comment:
There's no response defined for `discover` -- what is it for?

Reply:
Good catch, that part was a bit hastily added when I realised that `Discovery` is an integral part of the protocol orchestra. The intention for `discover` is to make a peer address known to a running daemon, similar to how `StreamDiscovery` works which can react to changes of the configured seed nodes in upstream at runtime. As the underlying operation is streaming in nature I'd assume the response would be a noop of thoughts, unless there is something else that is more indicative wrt the success of the request.

https://github.com/radicle-dev/radicle-link/blob/138813dba6d2cf377865ba163049982cb4c03c1c/daemon/src/config.rs#L93-L132

Reply:
Frankly, I don't like the terminology of `discover` much, but do think there is the need to manipulate the set of configured bootstrap nodes. If that is an agreeable statement that means it needs more scrutiny if these changes are persistent, if so where and how.

Reply:
> I don't like the terminology of discover much

Ya, this might get confusing as content discovery becomes a thing. Maybe it should be `peer-discovery`? Or a completely different term :)

Reply:
This should go away entirely. The idea of "discovery" is ill-defined, cf #629. The use case of query provider of piece-of-data, acquire piece-of-data can and should be hidden behind the `net::Peer` interface.

Reply:
To summarise no footgun should be handed out which interferes with the membership protocol as it has potential of incurring churn and undesired behaviour wrt. management of active/passive sets. In the short-term we accept a static set of boostrap nodes as discovery mechanism and mid-term expect the discovery interface to change to be pull based when the network stack has the need for a fresh node to connect to. Does that roughly sum up the state of affairs?

If that is the scenario, I'm happy to drop the discovery protocol from the document - one less thing to worry about - and capture follow up work around discovery in issue(s).

Reply:
Dropped the discovery protocol in 14c14a4 and created a tracking issue #723 to capture the follow-up work. Anything else we like to see captured for this?

Reply:
Closing this thread as #723 captures the conversation wrt. discovery changes.